Free BerkeleyEnvironment instances when not in use

ryanofsky opened the pull request:
Instead of adding BerkeleyEnvironment objects permanently to the g_dbenvs map, use reference counted shared pointers and remove map entries when the last BerkeleyEnvironment reference goes out of scope.

This change was requested by @TheBlueMatt and makes code that sets up mock databases cleaner. The mock database environment will now go out of scope and be reset on destruction so there is no need to call BerkeleyEnvironment::Reset() during wallet construction to clear out prior state.

This change does affect bitcoin behavior slightly. On startup, instead of same wallet environments staying open throughout VerifyWallets() and OpenWallets() calls, VerifyWallets() will open and close an environment once for each wallet, and OpenWallets() will create its own environment(s) later.

promag commented on src/wallet/db.h:
```diff
@@ -86,7 +87,7 @@ class CDBEnv
 
 /** Get CDBEnv and database filename given a wallet path. */
-CDBEnv* GetWalletEnv(const fs::path& wallet_path, std::string& database_filename);
+std::shared_ptr<CDBEnv> GetWalletEnv(const fs::path& wallet_path, std::string& database_filename);
```
typedef `std::shared_ptr<CDBEnv>` something? CDBEnvRef for instance.

ryanofsky replied:
> typedef std::shared_ptr<CDBEnv> something? CDBEnvRef for instance

I don't think this is a good idea. I'd be reluctant to introduce yet another wallet database type, especially one that obfuscates what it represents and how it relates to CDBEnv. Use of ref type aliases has been justified in other places on the basis that where types are widely used, having aliases makes it easier to switch implementations without having to update a lot of existing code. But that justification doesn't apply here because CDBEnv is internal to db.h/db.cpp and not widely used.

promag replied:
:+1:

promag commented on src/wallet/db.cpp:
```diff
@@ -121,6 +124,7 @@ BerkeleyEnvironment::BerkeleyEnvironment(const fs::path& dir_path) : strPath(dir
 
 BerkeleyEnvironment::~BerkeleyEnvironment()
 {
+    g_dbenvs.erase(strPath);
```
Missing `cs_db` lock?

promag replied:
Could you change to `std::string const strPath;` in `BerkeleyEnvironment`?

promag replied:
Nit:
```
int count = g_dbenvs.erase(strPath);
assert(count == 1);
```

sipa commented on src/wallet/db.cpp:
```diff
@@ -72,11 +72,14 @@ BerkeleyEnvironment* GetWalletEnv(const fs::path& wallet_path, std::string& data
+        inserted.first->second = env;
+        return env;
+    } else {
+        return inserted.first->second.lock();
```
I may be missing the logic here, but how is it guaranteed that the `shared_ptr` hasn't expired yet?

promag replied:
It's valid until `~BerkeleyEnvironment()` which means `shared_ptr` just expired, and in `~BerkeleyEnvironment()` the map is erased.

sipa replied:
Ah, of course, makes sense.

sipa commented:
Concept ACK

ryanofsky committed:
Free BerkeleyEnvironment instances when not in use

Instead of adding BerkeleyEnvironment objects permanently to the g_dbenvs map,
use reference counted shared pointers and remove map entries when the last
BerkeleyEnvironment reference goes out of scope.

This change was requested by Matt Corallo <git@bluematt.me> and makes code that
sets up mock databases cleaner. The mock database environment will now go out
of scope and be reset on destruction so there is no need to call
BerkeleyEnvironment::Reset() during wallet construction to clear out prior
state.

This change does affect bitcoin behavior slightly. On startup, instead of same
wallet environments staying open throughout VerifyWallets() and OpenWallets()
calls, VerifyWallets() will open and close an environment once for each wallet,
and OpenWallets() will create its own environment(s) later.

promag commented:
@ryanofsky ping.

PierreRochard committed:
Trivial: add doxygen-compatible comments relating to BerkeleyEnvironment

PierreRochard committed:
Tests: add unit tests for GetWalletEnv

PierreRochard reviewed (approved):
Tested ACK a35b4ef797783596dff9ca24fc46789a84e90014

If it's useful I wrote a few unit tests https://github.com/PierreRochard/bitcoin/commit/fc47a8843b03718529c402fd42cabb7a0834a9ff and code comments https://github.com/PierreRochard/bitcoin/commit/ea7f8291dab051cceb2751ef40f7d2a0ee67f95c

PierreRochard commented:
One more thing: PR title and commit message should be updated to say BerkeleyEnvironment

practicalswift commented on src/wallet/db.cpp:
```diff
@@ -72,11 +79,14 @@ BerkeleyEnvironment* GetWalletEnv(const fs::path& wallet_path, std::string& data
+        auto env = std::make_shared<BerkeleyEnvironment>(env_directory.string());
+        inserted.first->second = env;
+        return env;
+    } else {
```
Drop this redundant `else` and move the `return` the outer scope?

promag commented on src/wallet/db.cpp:
```diff
@@ -72,11 +79,13 @@ BerkeleyEnvironment* GetWalletEnv(const fs::path& wallet_path, std::string& data
-    // pointers instead of objects, anyway.
-    return &g_dbenvs.emplace(std::piecewise_construct, std::forward_as_tuple(env_directory.string()), std::forward_as_tuple(env_directory)).first->second;
+    auto inserted = g_dbenvs.emplace(env_directory.string(), std::weak_ptr<BerkeleyEnvironment>());
+    if (inserted.second) {
```
I think this condition should be replaced with:
```
auto env = inserted.first->second.lock();
if (!env) {
    // ...
```
which prevents returning null `std::shared_ptr<BerkeleyEnvironment>`.

practicalswift commented on src/wallet/wallet.cpp:
```diff
@@ -3851,6 +3851,9 @@ bool CWallet::Verify(std::string wallet_file, bool salvage_wallet, std::string&
     }
 
+    // Keep same database environment instance across Verify/Recover calls below.
+    std::unique_ptr<WalletDatabase> database = WalletDatabase::Create(wallet_path);
```
This declaration hides the member `CWallet::database`. Is that intentional? If not, use another name? :-)

MeshCollider commented:
utACK https://github.com/bitcoin/bitcoin/pull/11911/commits/530efe9813fc1fc874c7b612cfd46cc00beb4bf3 modulo @promag's comments above, which need addressing

ken2812221 reviewed (approved):
utACK 14bc2a17dd03ccd89f65a302328763ff22c710c2. Could remove `FUNCTIONAL_TESTS_CONFIG="--exclude wallet_multiwallet.py"` from travis option

practicalswift commented:
@ryanofsky Can you try without `FUNCTIONAL_TESTS_CONFIG`? :-)

MarcoFalke commented:
@ryanofsky Are you still working on this?

ryanofsky commented:
> @ryanofsky Are you still working on this?

Yeah, I went back and implemented all the suggestions, but for some reason the assert in https://github.com/bitcoin/bitcoin/pull/11911#discussion_r185654925 is constantly triggering, so I need to debug.

promag commented:
Tested ACK 14bc2a17dd while implementing #15297.

laanwj commented:
utACK 14bc2a17dd03ccd89f65a302328763ff22c710c2
